Feat: Add support for text mints

kittypurrry opened the pull request:
This PR adds support for ASCII text mints. There are two new options added to the `MintForm` component - _This is a text mint_ checkbox and a _monospace font required_ checkbox.

For typed mints, the cover and thumbnail images are automatically generated based on the text input and the font selected. The description will also automatically be replaced by the text mint input.

**Minting Form**
The textarea renders the monospace (IBM Plex Mono) / non-monospace font (Source Sans Pro) depending on the _monospace font required_ checkbox value.

**_Monospace:_**
![image](https://github.com/teia-community/teia-ui/assets/4899571/3e51caf1-f967-4ba4-a8df-17440764dfd8)

**_Non-monospace:_**
![image](https://github.com/teia-community/teia-ui/assets/4899571/3afc293d-97c8-4008-92c8-2d1ca6bad7d3)


**Preview / Token Page**
_Example: /objkt/850488_

The preview component displays the typed art, in either a monospace font a non-monospace font (Source Sans Pro). If Monospace is selected, a 'monospace' tag is automatically added to ensure that it renders properly on OBJKT; and viceversa.

![image](https://github.com/teia-community/teia-ui/assets/4899571/d992c8ae-1481-4397-8f60-3f02ba85c2a1)

**Feed Page**

Text NFTs are rendered as their cover/thumbnail images on the feed listing page, similar to how OBJKT does it. The actual text mint is only rendered as in the individual token page. 

![image](https://github.com/teia-community/teia-ui/assets/4899571/799ab0e9-8247-48ca-8f4f-4391835a5a78)

Notes
- Source Sans Pro was selected for text mints as it's the default font (non-mono) on OBJKT. This ensures text are encoded correctly and displayed as intended on OBJKT and Teia. 
- IBM Plex Mono was selected for monospace text mints as it was the font used by typed.art. Currently, it seems that OBJKT does not provide much monospace support for text mints, and a feature request has been raised here: https://roadmap.objkt.com/b/software-feature-request/use-a-monospaced-font-for-text-tokens/. As a result, do expect that some of the monospace typed art would appear broken on OBJKT. However, Teia has taken care of it by rendering the right fonts for your monospace typed art. See an example minted type mint on OBJKT: https://objkt.com/tokens/hicetnunc/850488

--

_Update [21/04]:_
- UltimateTextToImage package has been uninstalled, replace by using node-canvas as done in typed.art._ 
- Moved 'This is a text mint' checkbox to be before description, so that when checked, the description box is automatically hidden and replaced by the text mint input itself.

**Non-text mint**
<img width="810" alt="image" src="https://github.com/teia-community/teia-ui/assets/4899571/d2834b75-a08b-4ccc-9ab7-20ffd8279323">

**Text mint checked**
<img width="810" alt="image" src="https://github.com/teia-community/teia-ui/assets/4899571/cd716d33-6a49-4a70-80c1-b4fab6276b11">

_Update [05/05]:_
- Change Mono from IBM Plex Mono to Iosevka

kittypurrry committed:
feat: Add support for typed based nfts

- Create new textarea input component for typed based nfts that renders the correct fonts to be displayed correctly on OBJKT
- Modify existing FormFields to include new select options for user to specify if it's a typed nft and whether it requires monospace font
- Auto generate cover image based on typed input

kittypurrry committed:
Some updates to display token's displayuri instead of artifact

kittypurrry committed:
Update Monotype art to use 'IBM Mono Plex' to be consistent with typed.art

kittypurrry committed:
Remove monaco

kittypurrry committed:
Merge branch 'teia-community:main' into feat/typed-nft

kittypurrry committed:
Fix lint errors

kittypurrry committed:
Merge branch 'teia-community:main' into feat/typed-nft

kittypurrry committed:
Fix some UI issues while minting - switching between preview and edit tabs

kittypurrry committed:
Minor UI updates for very long text mints

kittypurrry committed:
Some minor adjustments

- Description will be automatially be the typed input; hide description field
- Rename Monospace checkbox field

kittypurrry committed:
Replace UltimateTextToImage with canvas solution

kittypurrry committed:
Merge branch 'main' into feat/typed-nft

melMass committed:
chore: ♻️ apply format

melMass committed:
fix: ✨ minor edits

melMass commented:
Thanks, almost good for me, besides the comment I just made I have a few questions:
- Why not use ioveska for `.mono`?
- I didn't follow the discussions that lead to this PR so I just want to be sure it was talked about with someone at OBJKT.com?

We can clean it up more in a separate PR, but fI think it's mostly fine for now if the metadata are good for everyone

melMass commented on src/utils/mint.ts:
```diff
@@ -114,3 +115,131 @@ export const generateCoverAndThumbnail = async (
+      }
+    }
+  })
+}
```
Can you simplify/refactor this a bit for clarity? I moved the text.length check first to un-nest it, changed ts types and use only async/await but I'm still not fully getting the logic of it.

kittypurrry committed:
Fix font issue for generated thumbnail and changed mono font to Iosevka

melMass commented:
Regarding the metadata, it follows typedart logic (minus the on chain text) so it's actually fine!

kittypurrry committed:
Add comments and refactor thumbnail generation a bit

kittypurrry committed:
Update char length before ellipsis since text size can accomodate for longer lengths

melMass committed:
Merge branch 'main' into feat/typed-nft

melMass committed:
style: 💎 fix various styles issues

Not coming from that PR directly but revealed by it.
Also adds ts,tsx and json to the linters.

melMass commented:
When checking the generated previews I get weird results:

https://cache.teia.rocks/ipfs/QmYdXkZM2nCX38gwrdsD21VG1Vq2P4eur1DgDMu3aLocWZ
https://cache.teia.rocks/ipfs/QmXnZrk9u4KUR9bWNem4oQQS6VpuCuS2FXFDosRqExAeww

Another thing is that it's currently not handling single long lines and I'm not sure what we want to happen in these cases:

**input text**
> In the realm of decentralization, power disperses like scattered light,freedom blooms in every node, no central might.Trust in the collective, no single hand's sway, innovation thrives in this distributed array. From governance to finance, it's a paradigm shift, where autonomy reigns, and control takes its lift.

![image](https://github.com/teia-community/teia-ui/assets/7041726/1cda03a0-976a-4490-ad5b-a2230e89cec3)

melMass committed:
chore: 🧹 convert typed_art utils to ts

melMass committed:
refactor: 📦 image generation logic

Mainly simplifying code, reusing variables.

This also adds a preview of the generation to the mint
form directly.

melMass committed:
fix: 🐛 issues reported by CI

melMass committed:
fix: preview generation

melMass committed:
chore: 🧹 prefer const

Zir0h committed:
hide scrollbars and add text feed to the menu

melMass commented:
Thanks Ziroh
There is one edge case that i think needs testing (cc @merchantcoppola) very long text (for the css thing that you just added)

Zir0h committed:
disable description in preview cards for text/plain tokens

Zir0h committed:
use overflow:scroll

Zir0h committed:
mess around with overflow / whitespace

Zir0h committed:
Merge branch 'main' into feat/typed-nft

Zir0h committed:
added padding
